pkg/ovs/ovn-nbctl.go

```diff
@@ -2178,3 +2189,281 @@ func (c Client) SetPolicyRouteExternalIds(priority int32, match string, nameIpMa
+	cmds := []string{"--format=csv", "--no-heading", "--data=bare", "--columns=_uuid,cidr,external_ids,options", "find", "dhcp_options"}
+	if needVendorFilter {
+		cmds = append(cmds, fmt.Sprintf("external_ids:vendor=%s", util.CniTypeName))
+	}
+	if len(ls) != 0 {
+		cmds = append(cmds, fmt.Sprintf("external_ids:ls=%s", ls))
+	}
+	if len(protocol) != 0 && protocol != kubeovnv1.ProtocolDual {
+		cmds = append(cmds, fmt.Sprintf("external_ids:protocol=%s", protocol))
+	}
+
+	output, err := c.ovnNbCommand(cmds...)
+	if err != nil {
+		klog.Errorf("failed to find dhcp options, %v", err)
+		return nil, err
+	}
+	entries := strings.Split(output, "\n")
+	dhcpOptionsList := make([]dhcpOptions, 0, len(entries))
+	for _, entry := range strings.Split(output, "\n") {
+		if len(strings.Split(entry, ",")) == 4 {
+			t := strings.Split(entry, ",")
+
+			externalIdsMap := map[string]string{}
+			for _, ex := range strings.Split(t[2], " ") {
+				ids := strings.Split(strings.TrimSpace(ex), "=")
+				if len(ids) == 2 {
+					externalIdsMap[ids[0]] = ids[1]
+				}
+			}
+
+			optionsMap := map[string]string{}
+			for _, op := range strings.Split(t[3], " ") {
+				kv := strings.Split(strings.TrimSpace(op), "=")
+				if len(kv) == 2 {
+					optionsMap[kv[0]] = kv[1]
+				}
+			}
+
+			dhcpOptionsList = append(dhcpOptionsList,
+				dhcpOptions{UUID: strings.TrimSpace(t[0]), CIDR: strings.TrimSpace(t[1]), ExternalIds: externalIdsMap, options: optionsMap})
+		}
+	}
+	return dhcpOptionsList, nil
+}
+
+func (c *Client) createDHCPOptions(ls, cidr, optionsStr string) (dhcpOptionsUuid string, err error) {
+	klog.Infof("create dhcp options ls:%s, cidr:%s, optionStr:[%s]", ls, cidr, optionsStr)
+
+	protocol := util.CheckProtocol(cidr)
+	output, err := c.ovnNbCommand("create", "dhcp_options",
+		fmt.Sprintf("cidr=%s", strings.ReplaceAll(cidr, ":", "\\:")),
+		fmt.Sprintf("options=%s", strings.ReplaceAll(optionsStr, ":", "\\:")),
+		fmt.Sprintf("external_ids=ls=%s,protocol=%s,vendor=%s", ls, protocol, util.CniTypeName))
+	if err != nil {
+		klog.Errorf("create dhcp options %s for switch %s failed: %v", cidr, ls, err)
+		return "", err
+	}
+	dhcpOptionsUuid = strings.Split(output, "\n")[0]
+
+	return dhcpOptionsUuid, nil
+}
+
+func (c *Client) updateDHCPv4Options(ls, v4CIDR, v4Gateway, dhcpV4OptionsStr string) (dhcpV4OptionsUuid string, err error) {
+	dhcpV4OptionsStr = strings.ReplaceAll(dhcpV4OptionsStr, " ", "")
+	dhcpV4Options, err := c.ListDHCPOptions(true, ls, kubeovnv1.ProtocolIPv4)
+	if err != nil {
+		klog.Errorf("list dhcp options for switch %s protocol %s failed: %v", ls, kubeovnv1.ProtocolIPv4, err)
+		return "", err
+	}
+
+	if len(v4CIDR) > 0 {
+		if len(dhcpV4Options) == 0 {
+			// create
+			mac := util.GenerateMac()
+			if len(dhcpV4OptionsStr) == 0 {
+				// default dhcp v4 options
+				dhcpV4OptionsStr = fmt.Sprintf("lease_time=%d,router=%s,server_id=%s,server_mac=%s", 3600, v4Gateway, "169.254.0.254", mac)
```

Comment:
I check the ovn-nb manual, and it says 

>options : server_id: optional string
              The IP address for the DHCP server to use. This should be in the
              subnet  of  the  offered  IP.  This is also included in the DHCP
              offer as option 54, ``server identifier.’’ 

can the default value be the subnet gateway ip address?

Reply:
Both the subnet gateway ip and the Link-local address can work for the dhcp request. But the subnet gateway ip would be changed, so i think the Link-local address would be easy to config.